### ISSUE TITLE
ckeygen --changepass truncates private keys to zero bytes

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5894]
|<img alt="ltaylor.volks's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| ltaylor.volks reported|
|-|-|
|Trac ID|trac#5894|
|Type|defect|
|Created|2012-08-20 19:57:21Z|
|Branch|https://github.com/twisted/twisted/tree/ckeygen-changepass-5894|

`ckeygen --changepass` does not work and exhibits different failing behavior for unencrypted and encrypted keys.

 1. Unencrypted keys are truncated to zero bytes because the key is opened and written to in one call, which throws a TypeError (no data is written).
 2. Encrypted keys don't get destroyed, but the passphrase change also fails to complete due to an uncaught EncryptedKeyError. The user is never prompted for a passphrase and old/new passphrases provided on the command line are not used

**1. Unencrypted key - TypeError, nulls your private key**
```
$ ckeygen -f /Users/blah/.ssh/id_rsa -p
Enter new passphrase (empty for no passphrase): 
Enter same passphrase again: 
Traceback (most recent call last):
  File "/Users/blah/Develop/virtualenvs/lodgeprox/bin/ckeygen", line 7, in <module>
    execfile(__file__)
  File "/Users/blah/Develop/code/3rdparty/twisted-hg/bin/conch/ckeygen", line 15, in <module>
    run()
  File "/Users/blah/Develop/code/3rdparty/twisted-hg/twisted/conch/scripts/ckeygen.py", line 66, in run
    changePassPhrase(options)
  File "/Users/blah/Develop/code/3rdparty/twisted-hg/twisted/conch/scripts/ckeygen.py", line 144, in changePassPhrase
    keys.Key(key).toString(passphrase=options['newpass']))
TypeError: toString() got an unexpected keyword argument 'passphrase'
```

Fix:
 * The call to `toString` needs to be fixed to provide the passphrase using the correct arg
 * The private key should not be opened for writing until the key has been regenerated with the passphrase
 * Needs tests

**2. Encrypted key - EncryptedKeyError**
```
$ ckeygen -f /Users/blah/.ssh/id_rsa -p
Traceback (most recent call last):
  File "/Users/blah/Develop/virtualenvs/lodgeprox/bin/ckeygen", line 7, in <module>
    execfile(__file__)
  File "/Users/blah/Develop/code/3rdparty/twisted-hg/bin/conch/ckeygen", line 15, in <module>
    run()
  File "/Users/blah/Develop/code/3rdparty/twisted-hg/twisted/conch/scripts/ckeygen.py", line 66, in run
    changePassPhrase(options)
  File "/Users/blah/Develop/code/3rdparty/twisted-hg/twisted/conch/scripts/ckeygen.py", line 125, in changePassPhrase
    key = keys.Key.fromFile(options['filename']).keyObject
  File "/Users/blah/Develop/code/3rdparty/twisted-hg/twisted/conch/ssh/keys.py", line 66, in fromFile
    return Class.fromString(file(filename, 'rb').read(), type, passphrase)
  File "/Users/blah/Develop/code/3rdparty/twisted-hg/twisted/conch/ssh/keys.py", line 95, in fromString
    return method(data, passphrase)
  File "/Users/blah/Develop/code/3rdparty/twisted-hg/twisted/conch/ssh/keys.py", line 218, in _fromString_PRIVATE_OPENSSH
    raise EncryptedKeyError('encrypted key with no passphrase'
twisted.conch.ssh.keys.EncryptedKeyError: encrypted key with no passphrase
```

Fix:
 * Catch and deal with `EncryptedKeyError`
 * If user provides old and new passphrases, use them
 * Needs tests

Attachments:

* [5894-ckeygen.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/5e1/5e1ee51553b8621fd3f7e4f9599f0447962afa05/079eb6fcfbe90232b8cb32a92ecdb2a45e23dd5a.diff) (22321 bytes) - added by ltaylor.volks on 2012-08-21 00:54:26Z - Fix ckeygen --changepass; add tests
* [5894.bugfix](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/5e1/5e1ee51553b8621fd3f7e4f9599f0447962afa05/fbe2e36021620883fabc2211d26d144b83d5c545.bugfix) (143 bytes) - added by ltaylor.volks on 2012-08-21 00:58:39Z - Added topfile NEWS entry
* [5894-ckeygen.1.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/5e1/5e1ee51553b8621fd3f7e4f9599f0447962afa05/bebd43c242a707dc0ad5f604d99c81a65a15c88a.diff) (13051 bytes) - added by ltaylor.volks on 2012-08-29 16:26:16Z - Updated tests
* [5894-ckeygen.2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/5e1/5e1ee51553b8621fd3f7e4f9599f0447962afa05/26258ce0ab4a8736f1084925d98ba7113e85ae79.diff) (11417 bytes) - added by ltaylor.volks on 2012-09-22 18:01:16Z - Updated patch to apply cleanly against r35757

<details><summary>Searchable metadata</summary>

```
trac-id__5894 5894
type__defect defect
reporter__ltaylor_volks ltaylor.volks
priority__normal normal
milestone__ 
branch__branches_ckeygen_changepass_5894 branches/ckeygen-changepass-5894
branch_author__therve therve
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__ 
time__1345492641000000 1345492641000000
changetime__1372607958000000 1372607958000000
version__None None
owner__therve therve
cc__z3p
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="ltaylor.volks's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|ltaylor.volks commented|
|-|-|

Change passphrase issues fixed, tests added. Note that this patch includes cumulative changes from [#5889](https://github.com/twisted/twisted/issues/5889) and  [#5890](https://github.com/twisted/twisted/issues/5890) as the issues are related.

However there are intermittent failures in this new `twisted.conch.test.test_ckeygen.KeyGenTests.test_changePassPhrase`, specifically the test for providing a bad passphrase for an encrypted key.

Single runs of the test usually pass, but continuously running it will eventually bomb with a variety of errors that seem to stem from pyasn1.codec.ber.decoder() returning 'weird stuff'. I don't know if this is a problem with pyasn1 or the data being provided to it.

```
trial -u twisted.conch.test.test_ckeygen.KeyGenTests.test_changePassPhrase
...
Traceback (most recent call last):
  File "/Users/blah/Develop/code/3rdparty/twisted-hg/twisted/conch/test/test_ckeygen.py", line 172, in test_changePassPhrase
    {'filename':filename, 'pass':'wrongpassphrase'})
twisted.trial.unittest.FailTest: <type 'exceptions.AttributeError'> raised instead of SystemExit:
 Traceback (most recent call last):
  File "/Users/blah/Develop/code/3rdparty/twisted-hg/twisted/trial/unittest.py", line 1143, in _run
    utils.runWithWarningsSuppressed, self._getSuppress(), method)
  File "/Users/blah/Develop/code/3rdparty/twisted-hg/twisted/internet/defer.py", line 134, in maybeDeferred
    result = f(*args, **kw)
  File "/Users/blah/Develop/code/3rdparty/twisted-hg/twisted/internet/utils.py", line 191, in runWithWarningsSuppressed
    result = f(*a, **kw)
  File "/Users/blah/Develop/code/3rdparty/twisted-hg/twisted/conch/test/test_ckeygen.py", line 172, in test_changePassPhrase
    {'filename':filename, 'pass':'wrongpassphrase'})
--- <exception caught here> ---
  File "/Users/blah/Develop/code/3rdparty/twisted-hg/twisted/trial/unittest.py", line 243, in failUnlessRaises
    result = f(*args, **kwargs)
  File "/Users/blah/Develop/code/3rdparty/twisted-hg/twisted/conch/scripts/ckeygen.py", line 132, in changePassPhrase
    options['filename'], passphrase=options['pass']).keyObject
  File "/Users/blah/Develop/code/3rdparty/twisted-hg/twisted/conch/ssh/keys.py", line 67, in fromFile
    return Class.fromString(file(filename, 'rb').read(), type, passphrase)
  File "/Users/blah/Develop/code/3rdparty/twisted-hg/twisted/conch/ssh/keys.py", line 96, in fromString
    return method(data, passphrase)
  File "/Users/blah/Develop/code/3rdparty/twisted-hg/twisted/conch/ssh/keys.py", line 264, in _fromString_PRIVATE_OPENSSH
    if len(decodedKey) == 2:  # alternate RSA key
exceptions.AttributeError: EndOfOctets instance has no attribute '__len__'
```

Subsequent runs produce an identical traceback except for the most recent call, which varies in 2 patterns:

 1. `AttributeError` because the value returned from pyasn1.codec.ber.decoder() varies...sometimes it's an EndofOctets instance, sometimes a Boolean, sometimes an Integer, sometimes None
 2. `ValueError` because the value returned from pyasn1.codec.ber.decoder() contains invalid literals...many different values seem to be returned.

**1. Different types are returned from pyasn1.codec.ber.decoder()**
```
  File "/Users/blah/Develop/code/3rdparty/twisted-hg/twisted/conch/ssh/keys.py", line 264, in _fromString_PRIVATE_OPENSSH
    if len(decodedKey) == 2:  # alternate RSA key
exceptions.AttributeError: Boolean instance has no attribute '__len__'
```

**2. decoded value contains invalid literals**
```
...
  File "/Users/blah/Develop/code/3rdparty/twisted-hg/twisted/conch/ssh/keys.py", line 269, in _fromString_PRIVATE_OPENSSH
    n, e, d, p, q = [long(value) for value in decodedKey[1:6]]
exceptions.ValueError: invalid literal for long() with base 10: '\x87'
```

```
...
  File "/Users/blah/Develop/code/3rdparty/twisted-hg/twisted/conch/ssh/keys.py", line 269, in _fromString_PRIVATE_OPENSSH
    n, e, d, p, q = [long(value) for value in decodedKey[1:6]]
exceptions.ValueError: invalid literal for long() with base 10: '\xe0'
```

```
...
  File "/Users/blah/Develop/code/3rdparty/twisted-hg/twisted/conch/ssh/keys.py", line 269, in _fromString_PRIVATE_OPENSSH
    n, e, d, p, q = [long(value) for value in decodedKey[1:6]]
exceptions.ValueError: invalid literal for long() with base 10: '+'
```

**Comment by twisted-trac** [comment]
|<img alt="ltaylor.volks's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|ltaylor.volks commented|
|-|-|

Revised and simplified tests. Issues described above:[comment:2 ltaylor.volks] for previous patch have been fixed by using separate TestCases

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve commented|
|-|-|

(In [36277]) Branching to 'ckeygen-changepass-5894'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve set owner to ltaylor.volks|
|-|-|

Thanks for the great patch, and sorry for not giving you feedback earlier. There are still about 3 sys.exit calls not covered by tests, it'd be great if you could provide those as a patch against the branch I created. Everything else looked good.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve removed owner|
|-|-|

I made the changes myself, build results here: buildbot.twistedmatrix.com/boxes-supported?branch=/branches/ckeygen-changepass-5894

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph commented|
|-|-|

There were some spurious looking failures and the build was a little old so I just updated it.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_9"></a>|@tomprince set owner to @therve|
|-|-|

1.`_make_getpass`:
   1. The name doesn't follow coding standard.
   2. `@param passphrases:`?
   3. The code seems more complicated the necesarry: `return iter(pass).next`
 2. I don't see any tests for the first case described in the ticket.
   - Relatedly, given the errors described in the ticket, I'd be inclined to want more assertions about the file contents after various cases.
 3. `test_fromStringErrors`
   1. I think the comment `# trying to decrypt an unencrypted key` is backwards, which is relevant to the code changes. 
 4. Is there a reason
```
#!
sys.exit('Could not change passphrase: old passphrase error')
```
     doesn't include the message from the exception?
 5. Is there a reason for changing `options['filename']` to use `get`? 
    - Incidentally, it seems that the prompt suggests there is a default, but that doesn't seem to be implemented.

Please resubmit for review after addressing the above points.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve removed owner|
|-|-|

Replying to [tom.prince](#note_9):
>  1. `_make_getpass`:
>    1. The name doesn't follow coding standard.

I changed it, but it's somewhat borderline here, as it's a prefixed method name.

>    2. `@param passphrases:`?

Documented.

>    3. The code seems more complicated the necesarry: `return iter(pass).next`

I don't think that works.

>  2. I don't see any tests for the first case described in the ticket.

I'm pretty sure it's test_changePassphraseCreateError 

>    - Relatedly, given the errors described in the ticket, I'd be inclined to want more assertions about the file contents after various cases.

Added some.

>  3. `test_fromStringErrors`
>    1. I think the comment `# trying to decrypt an unencrypted key` is backwards, which is relevant to the code changes.

Somewhat, changed.

>  4. Is there a reason
> ```
> #!
> sys.exit('Could not change passphrase: old passphrase error')
> ```
>      doesn't include the message from the exception?

The error message can be pretty cryptic, I think the message here is a bit better.

>  5. Is there a reason for changing `options['filename']` to use `get`?

Probably not, reverted.

>     - Incidentally, it seems that the prompt suggests there is a default, but that doesn't seem to be implemented.

Right we can open a bug about that.

> Please resubmit for review after addressing the above points.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_11"></a>|@tomprince set owner to @therve|
|-|-|

>> 3. The code seems more complicated the necesarry: `return iter(pass).next`
> I don't think that works
How about:
```python
    passphrases = iter(passphrases)
    def fakeGetpass(_):
        return passphrases.next()
    return fakeGetpass
```
Using a generator and `.send` seems too clever.

Applying the following patch doesn't cause any tests to fail:
```
#!diff
--- twisted/conch/scripts/ckeygen.py
+++ twisted/conch/scripts/ckeygen.py
@@ -160,7 +160,7 @@ def changePassPhrase(options):
         sys.exit('Could not change passphrase: %s' % (e,))
 
     fd = open(options['filename'], 'w')
-    fd.write(newkeydata)
+    fd.write('')
     fd.close()
 
     print 'Your identification has been saved with the new passphrase.'
```

>> - Incidentally, it seems that the prompt suggests there is a default, but that doesn't seem to be implemented.
> Right we can open a bug about that.
Sure.

Please merge after addressing the above three points.

**Comment by twisted-trac** [comment]
|<img alt="ltaylor.volks's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_12"></a>|ltaylor.volks commented|
|-|-|


Replying to [tom.prince](#note_11):
> >> 3. The code seems more complicated the necesarry: `return iter(pass).next`
> > I don't think that works
> How about:
```python
    passphrases = iter(passphrases)
    def fakeGetpass(_):
        return passphrases.next()
    return fakeGetpass
```
> Using a generator and `.send` seems too clever.


Agreed. Your solution is simpler and works.


> Applying the following patch doesn't cause any tests to fail:
```
#!diff
--- twisted/conch/scripts/ckeygen.py
+++ twisted/conch/scripts/ckeygen.py
@@ -160,7 +160,7 @@ def changePassPhrase(options):
         sys.exit('Could not change passphrase: %s' % (e,))
 
     fd = open(options['filename'], 'w')
-    fd.write(newkeydata)
+    fd.write('')
     fd.close()
 
     print 'Your identification has been saved with the new passphrase.'
```


By the time `newkeydata` is written it has already successfully survived roundtrip generation as a string from a Key object and converted to a valid Key object. More specifically, if newkeydata is an empty string (returned from toString), the subsequent call to fromString will raise BadKeyError and ckeygen will exit. This is covered by `twisted.conch.test.test_keys.KeyTestCase.test_fromStringErrors`

Here's a patch to address the review comments. I added a test to address the newkeydata concern (test_changePassphraseEmptyStringError), but I don't really think it's necessary.

therve: Thanks for picking this up and running with it (and thanks for the reviews tom.prince)!

```
#!diff
Index: twisted/conch/test/test_ckeygen.py
===================================================================
--- twisted/conch/test/test_ckeygen.py  (revision 38984)
+++ twisted/conch/test/test_ckeygen.py  (working copy)
@@ -34,13 +34,10 @@
 
     @param passphrases: The list of passphrases returned, one per each call.
     """
-    def _getpass():
-        yield None
-        for phrase in passphrases:
-            yield phrase
-    getpassResult = _getpass()
-    next(getpassResult)
-    return getpassResult.send
+    passphrases = iter(passphrases)
+    def fakeGetpass(_):
+        return passphrases.next()
+    return fakeGetpass
 
 
 
@@ -284,6 +281,30 @@
         self.assertEqual(privateRSA_openssh, FilePath(filename).getContent())
 
 
+    def test_changePassphraseEmptyStringError(self):
+        """
+        L{changePassPhrase} doesn't modify the key file if toString
+        returns an empty string
+        """
+        filename = self.mktemp()
+        FilePath(filename).setContent(privateRSA_openssh)
+
+        def toString(*args, **kwargs):
+            return ''
+
+        self.patch(Key, 'toString', toString)
+
+        error = self.assertRaises(
+            SystemExit, changePassPhrase,
+            {'filename': filename,
+             'newpass': 'newencrypt'})
+
+        self.assertEqual(
+            "Could not change passphrase: cannot guess the type of ''", str(error))
+
+        self.assertEqual(privateRSA_openssh, FilePath(filename).getContent())
+
+
     def test_changePassphrasePublicKey(self):
         """
         L{changePassPhrase} exits when trying to change the passphrase on a
```

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_13"></a>|@therve commented|
|-|-|

(In [38998]) Apply the given patch

Refs [#5894](https://github.com/twisted/twisted/issues/5894)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_14"></a>|@therve set status to `closed`|
|-|-|

(In [38999]) Merge ckeygen-changepass-5894

Authors: ltaylor.volks, therve
Reviewers: tom.prince, therve
Fixes: [#5894](https://github.com/twisted/twisted/issues/5894)

Fix ckeygen --changepass behavior, so that it doesn't accidently delete
unencrypted keys or fail with encrypted ones.
